File: src/main/java/es/daw/springdiscord/testspringbot/bot/DiscordListener.java
package es.daw.springdiscord.testspringbot.bot;

import es.daw.springdiscord.testspringbot.interfaces.ICommand;
import lombok.NonNull;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Component
public class DiscordListener extends ListenerAdapter {

    private final Map<String, ICommand> comandos = new HashMap<>();

    @Autowired
    public void setComandos(List<ICommand> listaComandos) {
        for (ICommand cmd : listaComandos) {
            comandos.put(cmd.getName(), cmd);
        }
    }

    @Override
    public void onSlashCommandInteraction(@NonNull SlashCommandInteractionEvent event) {

        ICommand comando = comandos.get(event.getName());

        if (comando != null) {
            event.deferReply().queue();
            comando.execute(event);
        }


    }
}

File: src/main/kotlin/parallaxdiscordbot/bot/DiscordListener.kt
package es.daw.parallaxdiscordbot.bot

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.slf4j.LoggerFactory

class DiscordListener(
    commandList: List<ICommand>,
    private val scope: CoroutineScope
) : ListenerAdapter() {

    private val commandList = commandList.associateBy { it.name }
    private val logger = LoggerFactory.getLogger(DiscordListener::class.java)

    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
        val commandName = event.name

        logger.info("Executing command $commandName")

        scope.launch {
            val startTime = System.currentTimeMillis()
            try {
                commandList[commandName]?.let { command ->
                    command.execute(event)
                    val duration = System.currentTimeMillis() - startTime
                    logger.info("Command $commandName completed in $duration ms")
                }
            } catch (e: Exception) {
                logger.error("Error while executing command $commandName", e)
            }
        }
    }
}
